kernel/src/schema/diff.rs

```diff
@@ -0,0 +1,1739 @@
+        ]);
+
+        let diff = SchemaDiffArgs {
+            current: &current,
+            new: &new,
+        }
+        .compute_diff()
+        .unwrap();
+
+        // Should see: existing changed, existing_struct.old_name->new_name renamed, new_struct added
+        assert_eq!(diff.added_fields.len(), 1);
+        assert_eq!(diff.added_fields[0].path, ColumnName::new(["new_struct"]));
+
+        assert_eq!(diff.updated_fields.len(), 2);
+        let paths: HashSet<ColumnName> =
+            diff.updated_fields.iter().map(|u| u.path.clone()).collect();
+        assert!(paths.contains(&ColumnName::new(["existing"])));
+        assert!(paths.contains(&ColumnName::new(["existing_struct", "new_name"])));
+
+        // nested_field should NOT appear as added since new_struct is its ancestor
+    }
+
+    #[test]
+    fn test_nested_field_rename() {
+        let current = StructType::new_unchecked([create_field_with_id(
+            "user",
+            DataType::try_struct_type([create_field_with_id("name", DataType::STRING, false, 2)])
+                .unwrap(),
+            false,
+            1,
+        )]);
+
+        let new = StructType::new_unchecked([create_field_with_id(
+            "user",
+            DataType::try_struct_type([
+                create_field_with_id("full_name", DataType::STRING, false, 2), // Renamed!
+            ])
+            .unwrap(),
+            false,
+            1,
+        )]);
+
+        let diff = SchemaDiffArgs {
+            current: &current,
+            new: &new,
+        }
+        .compute_diff()
+        .unwrap();
+        assert_eq!(diff.updated_fields.len(), 1);
+
+        let update = &diff.updated_fields[0];
+        assert_eq!(update.path, ColumnName::new(["user", "full_name"]));
+        assert_eq!(update.change_type, FieldChangeType::Renamed);
+    }
+
+    #[test]
+    fn test_nested_field_added() {
+        let current = StructType::new_unchecked([create_field_with_id(
+            "user",
+            DataType::try_struct_type([create_field_with_id("name", DataType::STRING, false, 2)])
+                .unwrap(),
+            false,
+            1,
+        )]);
+
+        let new = StructType::new_unchecked([create_field_with_id(
+            "user",
+            DataType::try_struct_type([
+                create_field_with_id("name", DataType::STRING, false, 2),
+                create_field_with_id("age", DataType::INTEGER, true, 3), // Added!
+            ])
+            .unwrap(),
+            false,
+            1,
+        )]);
+
+        let diff = SchemaDiffArgs {
+            current: &current,
+            new: &new,
+        }
+        .compute_diff()
+        .unwrap();
+        assert_eq!(diff.added_fields.len(), 1);
+
+        let added = &diff.added_fields[0];
+        assert_eq!(added.path, ColumnName::new(["user", "age"]));
+        assert_eq!(added.field.name(), "age");
+    }
+
+    #[test]
+    fn test_array_element_struct_field_changes() {
```

Comment:
also test that we can detect doubly nested array: `array<array<int>> -> array<array<double>>`